source/common/network/io_socket_handle_impl.cc

```diff
@@ -253,39 +248,115 @@ Api::IoCallUint64Result IoSocketHandleImpl::recvmsg(Buffer::RawSlice* slices,
+  if (hdr.msg_controllen > 0) {
+    // Get overflow, local address from control message.
+    for (struct cmsghdr* cmsg = CMSG_FIRSTHDR(&hdr); cmsg != nullptr;
+         cmsg = CMSG_NXTHDR(&hdr, cmsg)) {
+      if (output.msg_[0].local_address_ == nullptr) {
+        Address::InstanceConstSharedPtr addr = maybeGetDstAddressFromHeader(*cmsg, self_port, fd_);
         if (addr != nullptr) {
           // This is a IP packet info message.
-          output.local_address_ = std::move(addr);
+          output.msg_[0].local_address_ = std::move(addr);
           continue;
         }
-      } catch (const EnvoyException& e) {
-        PANIC(fmt::format("Invalid destination address for fd: {}, error: {}", fd_, e.what()));
+      }
+      if (output.dropped_packets_ != nullptr) {
+        absl::optional<uint32_t> maybe_dropped = maybeGetPacketsDroppedFromHeader(*cmsg);
+        if (maybe_dropped) {
+          *output.dropped_packets_ = *maybe_dropped;
+        }
       }
     }
-    if (output.dropped_packets_ != nullptr) {
-      absl::optional<uint32_t> maybe_dropped = maybeGetPacketsDroppedFromHeader(*cmsg);
-      if (maybe_dropped) {
-        *output.dropped_packets_ = *maybe_dropped;
+  }
+  return sysCallResultToIoCallResult(result);
+}
+
+Api::IoCallUint64Result IoSocketHandleImpl::recvmmsg(RawSliceArrays& slices, uint32_t self_port,
+                                                     RecvMsgOutput& output) {
+  ASSERT(output.msg_.size() == slices.size());
+  if (slices.size() == 0) {
+    return sysCallResultToIoCallResult(Api::SysCallIntResult{0, EAGAIN});
+  }
+  const uint32_t num_packets_per_mmsg_call = slices.size();
+  absl::FixedArray<mmsghdr> mmsg_hdr(num_packets_per_mmsg_call);
+  absl::FixedArray<absl::FixedArray<struct iovec>> iovs(
+      num_packets_per_mmsg_call, absl::FixedArray<struct iovec>(slices[0].size()));
+  absl::FixedArray<sockaddr_storage> raw_addresses(num_packets_per_mmsg_call);
+  absl::FixedArray<absl::FixedArray<char>> cbufs(num_packets_per_mmsg_call,
+                                                 absl::FixedArray<char>(cmsg_space_));
+
+  for (uint32_t i = 0; i < num_packets_per_mmsg_call; ++i) {
+    memset(&raw_addresses[i], 0, sizeof(sockaddr_storage));
+    memset(cbufs[i].data(), 0, cbufs[i].size());
+
+    mmsg_hdr[i].msg_len = 0;
+
+    msghdr* hdr = &mmsg_hdr[i].msg_hdr;
+    hdr->msg_name = &raw_addresses[i];
+    hdr->msg_namelen = sizeof(sockaddr_storage);
+    ASSERT(slices[i].size() > 0);
+
+    for (size_t j = 0; j < slices[i].size(); ++j) {
+      iovs[i][j].iov_base = slices[i][j].mem_;
+      iovs[i][j].iov_len = slices[i][j].len_;
+    }
+    hdr->msg_iov = iovs[i].data();
+    hdr->msg_iovlen = slices[i].size();
+    hdr->msg_control = cbufs[i].data();
+    hdr->msg_controllen = cbufs[i].size();
+  }
+
+  const Api::SysCallIntResult result = Api::OsSysCallsSingleton::get().recvmmsg(
+      fd_, mmsg_hdr.data(), num_packets_per_mmsg_call, MSG_TRUNC | MSG_WAITFORONE, nullptr);
```

Comment:
What's the behavior of this if MSG_WAITFORONE isn't set? It's a non-blocking socket, so I would have expected the behavior described in the man page for this flag without it.

Reply:
recvmmsg without MSG_WAITFORONE will block till received the specified number of packets which is not that convenient to use in test. This only matters for client socket in tests as all listen sockets is set to non-blocking.